doc/source/development/rfc/rfc110_jit_compilation.rst

```diff
@@ -0,0 +1,391 @@
+         * @param[out] pbHasVeclib Pointer to a boolean to indicate if a math vector lib
+         *                         has been found, or nullptr if not useful.
+         * @return a std::function of signature FunctionSignature corresponding to the
+         *         entry point in the C code (may be invalid in case of error.)
+         * @since 3.13
+         */
+        template <typename FunctionSignature>
+        std::function<FunctionSignature> CPL_DLL GDALGetJITFunction(
+            const std::string &cCode, const std::string &functionName,
+            CSLConstList papszOptions = nullptr,
+            std::string *posDisassembledCode = nullptr, bool *pbHasVeclib = nullptr);
+
+  Note that we return a `std::function` and not a C callback because the
+  code compiled by LLVM is owned by LLVM objects, and so we need to capture them
+  in the lambda we return as the `std::function`, so as to keep the underlying
+  C callable code valid as long as the `std::function` instance is valid.
+
+  To be noted that the compilation process does not involve any access to on-disk files
+  (verified with the ``strace`` utility). Only in-memory buffers are created.
+
+  There is also a function to indicate if JIT capabilities are present:
+
+  .. code-block:: cpp
+
+        /** Return which JIT engines are available.
+         *
+         * At time of writing, the return value may be an empty vector or a vector
+         * with "LLVM".
+         *
+         * @since 3.13
+         */
+        std::vector<std::string> CPL_DLL GDALGetJITEngines();
+
+  That header file is installed, as it may be useful to other projects.
+
+* gcore/gdal_jit.cpp:
+
+  Implementation of the above header, using libLLVM and libclang-cpp C++
+  semi-public/semi-private/unstable APIs.
+  That implementation has been verified to build and work for all versions
+  from LLVM 14 up to LLVM 22dev. There are a few #ifdef to account for breaking
+  changes in the LLVM API, but not that much.
+
+* gcore/gdal_c_expr.h:
+
+  Definition of a C++ class ``GDAL_c_expr_node`` and a
+  ``std::unique_ptr<GDAL_c_expr_node> CPL_DLL GDAL_c_expr_compile(const char *expr)``
+  method that parses a C-like MuParser expression string and return the root node
+  of the abstract syntax tree (AST) of the expression.
+  The parser supports all MuParser operators and functions, as well as our
+  ``nan``, ``isnan``, ``isnodata`` custom extensions.
+
+* gcore/gdal_c_expr.cpp: Implementation of the above
+
+* gcore/gdal_c_expr.y: Bison grammar for a C-like MuParser expression
+
+* gcore/gdal_c_expr_parser.cpp/.h:
+
+  Files generated from above file, and included in :file:`gdal_c_expr.cpp`
+
+
+Modified files
+++++++++++++++
+
+The ``ExprPixelFunc`` function in `frmts/vrt/pixelfunctions.cpp`, which is
+the method used to evaluate MuParser/ExprTK expressions is modified.
+
+When the JIT is present and a MuParser expression is used with that function,
```

Comment:
I find it really confusing to use "MuParser" to refer to something that is not going to be processed by the MuParser library. What I understand this RFC to propose is a third expression dialect ("C") that has a high overlap with the current version of MuParser when used to evaluate 64-bit floats.

Reply:
> What I understand this RFC to propose is a third expression dialect ("C" ?) that has a high overlap with the current version of MuParser when used to evaluate 64-bit floats.

not really, as far as I can tell, it understands the MuParser expressions. It is not a different dialect. Just an alternate implementation

Reply:
There may be subtle differences now or in the future (for example, the muparser library release notes show a change in bitwise operator precedence in the last release.) I don't understand the advantage in using the name "MuParser" if the expression is not evaluated by muparser. Do we want to track this library over time and ensure that an alternate implementation remains compatible?

Reply:
> I don't understand the advantage in using the name "MuParser" if the expression is not evaluated by muparser.

I don't want users to have to learn a new flavor of expression, so yes the idea would be to stick with emulating MuParser, which is very close to a subset of C (the main difference, beyond extensions, is the '^' operator which means power in MuParser and not bitwise exclusive or). It doesn't look like its capabilities are evolving at a very fast pace.  People using gdal raster calc use MuParser expressions by default, so this would be transparent to them